File: C++ Code/Basics/Calculator/Number System.cpp
#include<iostream>
int rev(int a){
    int ans=0;
    while(a>0){
        ans=ans*10+a%10;
        a/=10;
    }
    return ans;
}
bool CheckBinary(int a){
    while(a>0){
        if(a%10>1)
        return false;
        a/=10;    
    }
    return true;
}
bool CheckOctal(int a){
    while(a>0){
        if(a%10>7)
        return false;
        a/=10;    
    }
    return true;
}
int btd(int a){        // Binary to demimal conversion
    int pow=1,ans=0;
    while(a>0){
        ans+=(a%10)*pow;
        pow*=2;
        a/=10;
    }
    return ans;
}
int dtb(int a){        // Decimal to binary conversion
    if(a%2!=0){
        int ans=0;
        while(a>0){
            ans= ans*10+a%2;
            a/=2;
        }
    return rev(ans);
    }
    else{
        int ans=1;
        while(a>0){
            a/=2;
            ans= ans*10+a%2;
        }
    return rev(ans)-1;
    }
}
int dto(int a){        // Decimal to octal conversion
    if(a%8!=0){
        int ans=0;
        while(a>0){
            ans= ans*10+a%8;
            a/=8;
        }
    return rev(ans);
    }
    else{
        int ans=1;
        while(a>0){
            a/=8;
            ans= ans*10+a%8;
        }
    return rev(ans)-1;
    }
}
int otd(int a){        // Octal to decimal conversion
    int pow=1,ans=0;
    while(a>0){
        ans+=(a%10)*pow;
        pow*=8;
        a/=10;
    }
    return ans;
} 
int main(){
    int x,a;
    std::cout<<"Press 1 for binary to decimal conversion"<<std::endl;
    std::cout<<"Press 2 for binary to octal conversion"<<std::endl;
    std::cout<<"Press 3 for decimal to bianry conversion"<<std::endl;
    std::cout<<"Press 4 for decimal to octal conversion"<<std::endl;
    std::cout<<"Press 5 for octal to binary conversion"<<std::endl;
    std::cout<<"Press 6 for octal to decimal conversion"<<std::endl;
    std::cin>>x;
    std::cout<<std::endl;
    switch(x){
        case 1:
            std::cout<<"Enter your binary number : ";
            std::cin>>a;
            if(CheckBinary(a))
                std::cout<<"Decimal form of "<<a<<" = "<<btd(a);
            else
                std::cout<<a<<" is not a binary number.";
            break;
        case 2:
            std::cout<<"Enter your binary number : ";
            std::cin>>a;
            if(CheckBinary(a))
                std::cout<<"Octal form of "<<a<<" = "<<dto(btd(a));
            else
                std::cout<<a<<" is not a binary number.";
            break;
        case 3:
            std::cout<<"Enter your decimal number : ";
            std::cin>>a;
            std::cout<<"Binary form of "<<a<<" is "<<dtb(a);
            break;
        case 4:
            std::cout<<"Enter your decimal number : ";
            std::cin>>a;
            std::cout<<"Octal form of "<<a<<" is "<<dto(a);
            break;
        case 5:
            std::cout<<"Enter your octal number : ";
            std::cin>>a;
            if(CheckOctal(a))
                std::cout<<"Binary form of "<<a<<" is "<<dtb(otd(a));
            else
                std::cout<<a<<" is not a octal number.";    
            break;
        case 6:
            std::cout<<"Enter your octal number : ";
            std::cin>>a;
            if(CheckOctal(a))
                std::cout<<"Decimal form of "<<a<<" is "<<otd(a);
            else
                std::cout<<a<<" is not a octal number.";
            break;
        default :
            std::cout<<"Invalid option";    
    }
    return 0;
}

File: Java Code/Basics/Calculator/NumberSystem.java
import java.util.*;

public class NumberSystem{
  public static boolean inValidBinary(int num){
    for(int i=num;i>0;i/=10){
      if(i%10>1)
        return true;
    }
    return false;
  }
  public static boolean inValidOctal(int num){
    for(int i=num;i>0;i/=10){
      if(i%10>7)
        return true;
    }
    return false;
  }
  public static int rev(int num){
    int ans=0;
    for(int i=num;i>0;i/=10)
      ans=ans*10+i%10;
    return ans;
  }
  public static int btd(int num){
    int ans=0;
    for(int i=num,j=1;i>0;i/=10,j*=2)
      ans+=(i%10)*j;
    return ans;
  }
  public static int otd(int num){
    int ans=0;
    for(int i=num,j=1;i>0;i/=10,j*=8)
      ans+=(i%10)*j;
    return ans;
  }
  public static int dtb(int num){
    int a=num;
    if(num%2==0)
      num++;
    int ans=0;
    for(int i=num;i>0;i/=2)
      ans=ans*10+i%2;
    ans=rev(ans);
    if(a%2==0)
      return ans-1;
    return ans;
  }
  public static int dto(int num){
    int a=num;
    if(num%8==0)
      num++;
    int ans=0;
    for(int i=num;i>0;i/=8)
      ans=ans*10+i%8;
    ans=rev(ans);
    if(a%8==0)
      return ans-1;
    return ans;
  }
  public static void main(String[] args){
    Scanner sc=new Scanner(System.in);
    System.out.println("Choose the input format :\n1 for Binary number\n2 for Octal number\n3 for Decimal number.");
    int x=sc.nextInt();
    int num;
    switch(x){
      case 1:
        System.out.print("Enter the binary number : ");
        num=sc.nextInt();
        if(inValidBinary(num)){
          System.out.println(num+" is not a valid binary number.");
          return ;
        }
        System.out.println("For "+num+" in :");
        System.out.println(" ->  in Octal : "+dto(btd(num)));
        System.out.println(" ->  in Decimal : "+btd(num));
        break;
      case 2:
        System.out.print("Enter the octal number : ");
        num=sc.nextInt();
        if(inValidOctal(num)){
          System.out.println(num+" is not a valid octal format");
          return ;
        }
        System.out.println("For "+num+" in :");
        System.out.println(" ->  in Binary : "+dtb(otd(num)));
        System.out.println(" ->  in Decimal : "+otd(num));
        break;
      case 3:
        System.out.print("Enter decimal number : ");
        num=sc.nextInt();
        System.out.println("For "+num+" in :");
        System.out.println(" ->  in Binary : "+dtb(num));
        System.out.println(" ->  in Octal : "+dto(num));
        break;
      default:
        System.out.println("Enter a valid option.");
    }
  }
}